python/ray/data/random_access_dataset.py

```diff
@@ -0,0 +1,262 @@
+    """A class that provides distributed, random access to a Dataset.
+
+    See: ``Dataset.to_random_access_dataset()``.
+    """
+
+    def __init__(
+        self,
+        dataset: "Dataset[T]",
+        key: str,
+        num_workers: int,
+    ):
+        """Construct a RandomAccessDataset (internal API).
+
+        The constructor is not part of the Datasets API. Use
+        ``dataset.to_random_access_dataset()`` to construct a RandomAccessDataset.
+        """
+        self._format = dataset._dataset_format()
+        if self._format not in ["arrow", "pandas"]:
+            raise ValueError("RandomAccessDataset only supports Arrow-format datasets.")
+
+        start = time.perf_counter()
+        logger.info("[setup] Indexing dataset by sort key.")
+        sorted_ds = dataset.sort(key)
+        get_bounds = cached_remote_fn(_get_bounds)
+        blocks = sorted_ds.get_internal_block_refs()
+
+        logger.info("[setup] Computing block range bounds.")
+        bounds = ray.get([get_bounds.remote(b, key, self._format) for b in blocks])
+        self._valid_blocks = []
+        self._lower_bound = None
+        self._upper_bounds = []
+        for i, b in enumerate(bounds):
+            if b:
+                self._valid_blocks.append(blocks[i])
+                if self._lower_bound is None:
+                    self._lower_bound = b[0]
+                self._upper_bounds.append(b[1])
+
+        logger.info("[setup] Creating {} random access workers.".format(num_workers))
+        self._workers = [
+            _RandomAccessWorker.options(scheduling_strategy="SPREAD").remote(
+                key, self._format
+            )
+            for _ in range(num_workers)
+        ]
+        (
+            self._block_to_workers_map,
+            self._worker_to_blocks_map,
+        ) = self._compute_block_to_worker_assignments()
+
+        logger.info(
+            "[setup] Worker to blocks assignment: {}".format(self._worker_to_blocks_map)
+        )
+        ray.get(
+            [
+                w.assign_blocks.remote(
+                    {i: self._valid_blocks[i] for i in self._worker_to_blocks_map[w]}
+                )
+                for w in self._workers
+            ]
+        )
+
+        logger.info("[setup] Finished assigning blocks to workers.")
+        self._build_time = time.perf_counter() - start
+
+    def _compute_block_to_worker_assignments(self):
+        # Return values.
+        blocks_to_workers: dict[int, List["ray.ActorHandle"]] = defaultdict(list)
+        workers_to_blocks: dict["ray.ActorHandle", List[int]] = defaultdict(list)
+
+        # Aux data structures.
+        loc_to_workers: dict[str, List["ray.ActorHandle"]] = defaultdict(list)
+        locs = ray.get([w.ping.remote() for w in self._workers])
+        for i, loc in enumerate(locs):
+            loc_to_workers[loc].append(self._workers[i])
+        block_locs = ray.experimental.get_object_locations(self._valid_blocks)
```

Comment:
I was thinking about that, but it depends on the node pinning resources. For now, this strategy is safer since it avoids actors hanging forever on node loss.